File: BezierCurve/Bernstein.m
% Generated on: 191211
% Last modification: 191211
% Author: Suwon Lee from Seoul National University

classdef Bernstein
    properties
        coefficients
    end
    methods
        % Constructor
        function obj = Bernstein(varargin) 
            if ~isempty(varargin)
                obj.coefficients = varargin{1};
            end
        end
        function value = feval(obj,x)
            n = numel(obj.coefficients)-1;
            value = zeros(1,numel(x));;
            for ii = 1:n+1
                v = ii-1;
                value = value + obj.basis(n,v,x).*obj.coefficients(ii);
            end
        end
        function value = derivative(obj,x)
            n = numel(obj.coefficients)-1;
            value = zeros(1,numel(x));;
            for ii = 1:n+1
                v = ii-1;
                value = value + obj.basisPthDerivative(n,v,1,x).*obj.coefficients(ii);
            end
        end
        function value = Pthderivative(obj,x,P)
            n = numel(obj.coefficients)-1;
            value = zeros(1,numel(x));;
            for ii = 1:n+1
                v = ii-1;
                value = value + obj.basisPthDerivative(n,v,P,x).*obj.coefficients(ii);
            end
        end
    end
    methods (Static)
        function Bin = basis(n,i,x)
            Bin = nchoosek(n,i).*x.^i.*(1-x).^(n-i);
        end
        function DpBin = basisPthDerivative(n,i,p,x)
            DpBin_ = 0;
            for k = max(0,i+p-n):min(i,p)
                term = (-1)^(k+p).*nchoosek(p,k).*bernstein.basis(n-p,i-k,x);
                DpBin_ = DpBin_ + term;
            end
            DpBin = factorial(n)/factorial(n-p)*DpBin_;
        end
    end
end

File: pyBezier/bernstein.py
# Generated on: 2021-08-13 
# Author: Suwon Lee from Samsung Electronics

import numpy as np
import scipy
from scipy.special import comb

class Bernstein:
    def __init__(self,coef):
        self.coefficients = coef
        self.n = np.size(coef) - 1

    def basis(self,n,i,x):
        return comb(n,i)*x**i*(1-x)**(n-i)

    def basisPthDerivative(self,n,i,p,x):
        DpBin = 0
        for k in range(max(0,i+p-n),max(i,p)+1):
            term = (-1)**(k+p)*comb(p,k)*self.basis(n-p,i-k,x)
            DpBin = DpBin + term
        return scipy.special.factorial(n)/scipy.special.factorial(n-p)*DpBin

    def feval(self,x):
        n = self.n
        nx = np.size(x)
        value = np.zeros(nx)
        for i in range(1,n+1+1):
            v = i-1
            value = value + self.basis(n,v,x)*self.coefficients[v]
        return value

    def derivative(self,x):
        n = self.n
        nx = np.size(x)
        value = np.zeros(nx)
        for i in range(1,n+1+1):
            v = i-1
            value = value + self.basisPthDerivative(n,v,1,x)*self.coefficients[v]
        return value

    def Pthderivative(self,x,P):
        n = self.n
        nx = np.size(x)
        value = np.zeros(nx)
        for i in range(1,n+1+1):
            v = i-1
            value = value + self.basisPthDerivative(n,v,P,x)*self.coefficients[v]
        return value
